This notebook contains the following:

01) Renaming columns

02) Identifying the key column to merge dataframes

03) Addressing missing values and some PII (personal idenfiable information)

04) Merging dataframes and checking results


01. Starting new notebook for task 4.9
First notebook

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

02. Establishing path and importing dataframe

In [2]:
# Creating a path
path = r'C:\Users\Paola\Desktop\Shaul\Data\CF\August-2023 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Paola\\Desktop\\Shaul\\Data\\CF\\August-2023 Instacart Basket Analysis'

In [4]:
# Importing 'Customers' dataframe
dfcus = pd.read_pickle(os.path.join(path, '2-Data', 'Original Data', 'customers.pkl'))

In [5]:
# Checking df shape
dfcus.shape

(206209, 10)

In [8]:
# Checking Content and columns
dfcus.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
# Checking cells data types
dfcus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


03. Data wrangling procedures

In [10]:
# Renaming/Standardising column names
dfcus.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [11]:
dfcus.rename(columns = {'n_dependants' : 'dependants'}, inplace = True)

In [12]:
dfcus.rename(columns = {'STATE' : 'state'}, inplace = True)

In [13]:
dfcus.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [14]:
dfcus.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [15]:
dfcus.rename(columns = {'fam_status' : 'civil_status'}, inplace = True)

In [16]:
dfcus.rename(columns = {'Age' : 'age'}, inplace = True)

In [17]:
# Checking columns
dfcus.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants civil_status  income  
0           3      married  165665  
1           0       single   59285  
2           2      married   99568  
3           0       single   42049  
4           1      married   40374

In [18]:
dfcus.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'dependants', 'civil_status', 'income'],
      dtype='object')

User_id is the key to combine the data. the columns (name/last name) are necessary for identification. Gender, age and civil status are great for targeting specific customers. Dependants and income give an insight into the fact whether they are low or high spenders. However, the date_joined column is unnecessary: we already have a column with flags for new, regular and loyal customers.

In [19]:
# Dropping the 'date_joined' column
dfcus = dfcus.drop(columns = ['date_joined'])

04. Data quality and consistency

In [20]:
# Working on missing values
dfcus.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
dependants          0
civil_status        0
income              0
dtype: int64

In [22]:
# Creating a dataframe to find out the NaN entries
dfnan = dfcus[dfcus['first_name'].isnull() == True]

In [23]:
dfnan

user_id first_name last_name  gender         state  age  dependants  \
53        76659        NaN   Gilbert    Male      Colorado   26           2   
73        13738        NaN     Frost  Female     Louisiana   39           0   
82        89996        NaN    Dawson  Female        Oregon   52           3   
99        96166        NaN   Oconnor    Male      Oklahoma   51           1   
105       29778        NaN    Dawson  Female          Utah   63           3   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28           3   
206044   200799        NaN  Copeland  Female        Hawaii   52           2   
206090   167394        NaN     Frost  Female        Hawaii   61           1   
206162   187532        NaN     Floyd  Female    California   39           0   
206171   116898        NaN   Delgado  Female      Colorado   23           2   

       civil_status  income  
53          married   41709  
73           single   82518  
82          married  117099  
99          married  155673  
105         married  151819  
...             ...     ...  
206038      married   87783  
206044      married  108488  
206090      married   45275  
206162       single   56325  
206171      married   59222  

[11259 rows x 9 columns]

There are 11259 missing values in the 'first_name' column. In one hand, since this data is from within the organisation (probably coming from an OLTP), it is reliable. This must be the result of incorrect handling by the person in charge. I would go back to the client and ask about this. It would be possible to look for which customers the missing values are, and replace them accordingly. However, for the sake of analysis and time consumption, it is easier to drop the columns 'first_name' and 'last_name', as they do not contribute much for the analysis, and in case i would need something about a specific customer, I could go and filter the data, and replace those missing values.

In [24]:
# Dropping the abovementioned columns
dfcus = dfcus.drop(columns = ['first_name', 'last_name'])

In [25]:
# Checking dropped columns
dfcus.columns

Index(['user_id', 'gender', 'state', 'age', 'dependants', 'civil_status',
       'income'],
      dtype='object')

In [26]:
# Finding duplicates
dfdups = dfcus[dfcus.duplicated()]

In [27]:
dfdups

Empty DataFrame
Columns: [user_id, gender, state, age, dependants, civil_status, income]
Index: []

No duplicates were found

In [28]:
# Checking dfcus for mixing data
for col in dfcus.columns.tolist():
  weird = (dfcus[[col]].applymap(type) != dfcus[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (dfcus[weird]) > 0:
    print (col)

No mix data returned.

In [29]:
dfcus.head()

user_id  gender       state  age  dependants civil_status  income
0    26711  Female    Missouri   48           3      married  165665
1    33890  Female  New Mexico   36           0       single   59285
2    65803    Male       Idaho   35           2      married   99568
3   125935  Female        Iowa   40           0       single   42049
4   130797  Female    Maryland   26           1      married   40374

In [30]:
# Confirming missing values => no missing values found.
dfcus.isnull().sum()

user_id         0
gender          0
state           0
age             0
dependants      0
civil_status    0
income          0
dtype: int64

In [31]:
dfcus.shape

(206209, 7)

05. Merging dataframes
Importing readied 'Order/products' dataframe, and merging it with the new customers dataframe

In [28]:
# Importing 'orders_products_readyvariables.pkl' data set
dfprior = pd.read_pickle(os.path.join(path, '2-Data', 'Prepared Data', 'orders_products_readyvariables.pkl'))

In [29]:
# Merging both dataframes 
dfmerge = dfprior.merge(dfcus, on = ['user_id'])

In [30]:
dfmerge.shape

(32435059, 28)

In [32]:
dfmerge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_range_loc', 'Busiest days',
       'Busiest hours', 'max_order', 'loyalty_flag', 'avg_prices',
       'spending_flag', 'median_prior_order', 'frequency_flag', 'gender',
       'state', 'age', 'dependants', 'civil_status', 'income'],
      dtype='object')

In [33]:
# Confirming missing values => no missing values found.
dfmerge.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_prior_order        0
product_id                    0
add_to_cart_order             0
reordered                     0
product_name              30200
aisle_id                  30200
department_id             30200
prices                    30200
price_range_loc           30200
Busiest days                  0
Busiest hours                 0
max_order                     0
loyalty_flag                  0
avg_prices                    0
spending_flag                 0
median_prior_order            0
frequency_flag                0
gender                        0
state                         0
age                           0
dependants                    0
civil_status                  0
income                        0
dtype: int64

There are 30200values missing from the abovementioned rows. However, as this is less than 0.01% of the data, I am saving the df as it is, due to time constraint since it does not affect the analysis. 
In normal circunstances, I would have looked into it further.

06. Exporting new data frame to pkl format

In [34]:
dfmerge.to_pickle(os.path.join(path, '2-Data','Prepared Data', 'orders_products_customers.pkl'))